### DISCUSSION TITLE
ModuleEvent `event_index` definition for 0,1,2?

### DISCUSSION THREAD

**Discussion by gftea** [aptos-labs/aptos-developer-discussions#232] [Questions] [answered]
### Describe your question in detail.

When exploring graphQl api for module events, what is the definition/meaning of event_index 0,1,2, how they are decided?
I run the hello world example, and try to query the module event, only `0` and `1` returns data, but `2` return null. I cannot find the definition of these index.

<img width="1206" alt="image" src="https://github.com/aptos-labs/aptos-developer-discussions/assets/1705787/994b4efd-cce0-44fc-a097-55d3e8d0da9c">

### What have you tried or looked at?  Or how can we reproduce the error?

yes, use api https://cloud.hasura.io/public/graphiql?endpoint=https://api.testnet.aptoslabs.com/v1/graphql
While I also see documents in https://aptos.dev/concepts/events/#module-events, which just mentioned index 0,1,2, but does not explain how the index is decided.

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

REST API

**Comment by Alexzander-Stone** [comment, marked as answer]
## Event index

The indices indicate the order of emitted events during the TX. After an event is emitted, the event index is incremented.

## Hello World example

For your [hello world example](https://explorer.aptoslabs.com/txn/1212123728/events?network=testnet), the `MessageChange` event was emitted first and the TX ended with the `FeeStatement` event. Only two events were emitted.

So, when you query against this TX, you'll see the first event (`MessageChange`) as being index 0, and the second/last event (`FeeStatement`) as being index 1.

## Other examples

Check out the [docs-supplied example](https://explorer.aptoslabs.com/txn/682252266/events?network=testnet) for a showcase of a TX with 4 events.
